### ISSUE TITLE
preserve-fds: file descriptor 4 is not available

### ISSUE THREAD

**Issue by edsantiago** [containers/podman#15943]
Been seeing this flake for the last week:
```
$ podman [options] run --preserve-fds 2 quay.io/libpod/alpine:latest
Error: file descriptor 4 is not available - the preserve-fds option requires that file descriptors must be passed
```
### Podman run [It] podman run --preserve-fds invalid fd 
* ubuntu-2204 : int podman ubuntu-2204 rootless host
  * PR #15907
    * [09-25 22:53](https://api.cirrus-ci.com/v1/artifact/task/5786743352852480/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--1)
    * [09-25 22:53](https://api.cirrus-ci.com/v1/artifact/task/5786743352852480/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--2)
    * [09-25 22:53](https://api.cirrus-ci.com/v1/artifact/task/5786743352852480/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--3)
  * PR #15873
    * [09-22 14:51](https://api.cirrus-ci.com/v1/artifact/task/5638581090779136/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--1)
    * [09-22 14:51](https://api.cirrus-ci.com/v1/artifact/task/5638581090779136/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--2)
    * [09-22 14:51](https://api.cirrus-ci.com/v1/artifact/task/5638581090779136/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--3)
  * PR #15856
    * [09-20 04:19](https://api.cirrus-ci.com/v1/artifact/task/4677629470048256/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--1)
    * [09-20 04:19](https://api.cirrus-ci.com/v1/artifact/task/4677629470048256/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--2)
    * [09-20 04:19](https://api.cirrus-ci.com/v1/artifact/task/4677629470048256/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--3)
  * PR #15841
    * [09-22 13:51](https://api.cirrus-ci.com/v1/artifact/task/4847799094214656/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--1)
    * [09-22 13:51](https://api.cirrus-ci.com/v1/artifact/task/4847799094214656/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--2)
    * [09-22 13:51](https://api.cirrus-ci.com/v1/artifact/task/4847799094214656/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--3)
  * PR #15833
    * [09-16 09:01](https://api.cirrus-ci.com/v1/artifact/task/5684027582578688/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--1)
    * [09-16 09:01](https://api.cirrus-ci.com/v1/artifact/task/5684027582578688/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--2)
    * [09-16 09:01](https://api.cirrus-ci.com/v1/artifact/task/5684027582578688/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--3)
  * PR #15131
    * [09-20 17:08](https://api.cirrus-ci.com/v1/artifact/task/5626415361032192/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--1)
    * [09-20 17:08](https://api.cirrus-ci.com/v1/artifact/task/5626415361032192/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--2)
    * [09-20 17:08](https://api.cirrus-ci.com/v1/artifact/task/5626415361032192/html/int-podman-ubuntu-2204-rootless-host.log.html#t--podman-run---preserve-fds-invalid-fd--3)
 

Seems interesting that it's only Ubuntu rootless.

**Comment by edsantiago** [comment]
@containers/podman-maintainers please help. This one is starting to trigger often. 

ISTM that something somewhere is now doing an `open()` without a corresponding `close()`, leaving `fd3` stuck open forever.

The first instance of the flake was seen in PR #15833 (903f551ae5dabf51204ecdf2a340bf1b0589e5db). That doesn't mean this PR is at fault (I see nothing in it that leaks fds), it just means this is the upper bound of commits to search. Given how frequent this flake is, I'm guessing the lower bound is at most two weeks before that, so, let's say 54873c1f5 (Aug 31). That's impossible for me to review, even if I limit it to `git diff 54873c1f5 903f55 -- test/e2e`. (But I tried anyway).

What I need is for someone to think back to early September and say "oh, yeah, I added this-or-that code and might have forgotten to `close` somewhere". Pretty please, can you do this? The longer we wait, the harder it will be to remember.

If we don't get this resolved, I'm just going to remove the `preserve-fd` test entirely.

**Comment by edsantiago** [comment]
My belief came from the test failure message:
```
Expected
   <string>: Error: file descriptor 4 is not available - the preserve-fds option requires that file descriptors must be passed
to contain substring
   <string>: file descriptor 3 is not available
```
...but I may be reading it completely wrong.

**Comment by mheon** [comment]
Ack, that sounds accurate. An extra FD is getting in occasionally that it altering the error message.

No chance it's leaking in from CI itself passing an extra FD in?

**Comment by mheon** [comment]
Sorry if I wasn't clear, but I concur with your assessment. It's an extra FD, not a missing FD as I originally assumed.

**Comment by mheon** [comment]
I am, however, a bit doubtful that we missed a `defer file.Close()` somewhere - I seem to recall us having a linter that catches such things. Will verify that after lunch.

**Comment by edsantiago** [comment]
I found one through tedious skimming: https://github.com/containers/podman/blob/23a30661ad390e3a52fd99de241baa9e8d179e09/libpod/container_internal_common.go#L1086
(two, actually, there are two instances in that same file). But that's podman, not the e2e tests, so I can't imagine why that would affect fds in the e2e test process.

**Comment by Luap99** [comment]
> I am, however, a bit doubtful that we missed a `defer file.Close()` somewhere - I seem to recall us having a linter that catches such things. Will verify that after lunch.

I don't think there is a linter for that.


We use os/exec to start podman in the tests, it should not matter how many fds we leak in the test since os/exec will always unset all fds.
https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/os/exec/exec.go;l=506-517
That makes me think the leak or bug must be somewhere in podman itself.


ALso the golang garbage collector will automatically close files once the variable is out of scope, since the timing changes with every run it would explain why it flakes only sometimes.

**Comment by edsantiago** [comment]
For posterity, because it looks like I'm abandoning this again: thinking that perhaps the root cause was some test that ran before the flaking one, I wrote a script to analyze my logs, looking for tests that precede the invalid fd` one in all cases. No luck: there are 73 such matches. Too many to really evaluate.

Among those, in 11 of the 16 log files, `podman run with restart-policy always restarts containers` is the test run immediately before the flake. I looked at that test, it doesn't strike me as having anything to do with filehandles.

Moving on.

**Comment by eriksjolund** [comment]
Just an idea:

The issue

*  #15927 
 
also deals with a missing file descriptor (an activated socket).

I found a bug in rootless_linux.c

* https://github.com/containers/podman/pull/16188 

that seems to fix #15927. 
Maybe #15943 is related to that bug as well?

**Comment by edsantiago** [comment]
We no longer test Ubuntu. If we ever do so again, and this resurfaces, then I'll reopen.
